### PR TITLE
add MTP for GLM OCR and other MROPEs

### PR DESCRIPTION
**Pull request by molbap**
[CI](https://transformers-ci.lor-e.huggingface.cool/d/pytest-observability-by-pr/pytest-observability-branch?var-pr=47462)

# What does this PR do?

Since MTP is added in https://github.com/huggingface/transformers/pull/46229 we can start adding it. 
MROPE makes it a bit harder and I had to hack a bit modeling_layers, but the rest is relatively clean. Improves TPS by ~+50% from eyeballing it, need to benchmark it a bit better but looks good!

Edit 1: did some benchmarkings. Side note, on some configs (a100) conv3d bf16 with bias is ~20000x slower than f32 no bias :grimacing: I used fusion mapping to use a `nn.Linear` instead, much better. With fusion mapping I get:

```md
  baseline: 445 tokens in 9.4s (47.6 tok/s, 445 forward calls, 1.00 tokens/round)
  mtp     : 445 tokens in 6.7s (65.9 tok/s, 225 forward calls, 1.98 tokens/round)
  speedup: 1.39x, outputs identical (445 tokens)
```
pretty good & cheap improvement. Also added a test.

### CODE REVIEW

**Comment by molbap** [comment]
cc @Cyrilvallez @zucchini-nlp FYI (wip)

**Comment by molbap** [comment]
run-slow: glm_ocr

File: src/transformers/conversion_mapping.py
Comment (zucchini-nlp) [inline src/transformers/conversion_mapping.py]: oke, i missed this when MTP was added so I'm curious - do we assume that all models follow certain naming conventions in MTP? Looks weird adding it all under here, prob I am missing background info

File: src/transformers/modeling_layers.py [hunk 1]
```diff
@@ -496,20 +496,28 @@ def forward(
         drafted_tokens = []
         loss = None
         for i, mtp_layer in enumerate(self.layers):
+            # MROPE models (e.g. Glm4v family) pack positions as [4, bs, seq]: plane 0 holds the text
+            # positions used for the masks and layers, planes 1: hold the rope positions
+            if position_ids is not None and position_ids.ndim == 3 and position_ids.shape[0] == 4:
+                text_position_ids = position_ids[0]
+                rope_position_ids = position_ids[1:]
```
Comment (zucchini-nlp) [inline src/transformers/modeling_layers.py]: hmm the first dim actually isn't always 4. We recently merged HunyuanVL and they apparently support different number of mrope-axis (ig 2 without temporal rotations)

https://github.com/huggingface/transformers/blob/2df80c4f5863192f1d4cc698bd4f1e1d254e14ba/src/transformers/models/hunyuan_vl/modeling_hunyuan_vl.py#L813-L823

File: tests/generation/test_utils.py [hunk 2]
```diff
@@ -2426,7 +2426,7 @@ def test_generate_with_mtp(self):
             keys_to_ignore_unexpected = model_class._keys_to_ignore_on_load_unexpected or []
             # If we don't have any mtp patterns, skip
             if not hasattr(config.get_text_config(), "num_mtp_layers") or not any(
-                "mtp" in x or re.search(r"layers\.\d+", x) is not None for x in keys_to_ignore_unexpected
+                "mtp" in x or re.search(r"layers\\?\.\d+", x) is not None for x in keys_to_ignore_unexpected
```
Comment (zucchini-nlp) [inline tests/generation/test_utils.py]: any reason for this?

**Review by zucchini-nlp — COMMENTED** [review]
Great addition, i have only one concern about the hard check on `ids.shape[0] == 4`. I think we can make this more general and check if the length is one more than number of rotation axis, wdyt?

File: tests/generation/test_utils.py (hunk 2 above)
Comment (molbap) [reply tests/generation/test_utils.py]: yeah, the layers in this checkpoint :grimacing: will see if we can do better

File: src/transformers/conversion_mapping.py
Comment (molbap) [reply src/transformers/conversion_mapping.py]: yeah it is a bit brittle at the moment. @Cyrilvallez I don't know what you think about this, but overall this was more for Inkling.

File: src/transformers/modeling_layers.py (hunk 1 above)
Comment (molbap) [reply src/transformers/modeling_layers.py]: AH dang. of course. Checking

File: src/transformers/conversion_mapping.py
Comment (Cyrilvallez) [reply src/transformers/conversion_mapping.py]: Yeah, the issue is that the model class will always be `MtpModel`, but different actual model want different naming conversions for their mtp layers...
I def put a mental note that we should start having a subdir for main_model_type for the mtp conversions to avoid creating collisions when we add more - this could be the opportunity

File: src/transformers/generation/candidate_generator.py [hunk 3]
```diff
@@ -1507,15 +1507,15 @@ def get_candidates(
 
             # We need the last invalidated tokens, as well as the new one in a single passcfor efficiency
             mtp_input_ids = input_ids[:, -num_last_main_model_tokens - self.num_mtp_layers :]
-            mtp_position_ids = model_kwargs["position_ids"][:, -num_last_main_model_tokens - self.num_mtp_layers :]
+            mtp_position_ids = model_kwargs["position_ids"][..., -num_last_main_model_tokens - self.num_mtp_layers :]
```
Comment (Cyrilvallez) [inline src/transformers/generation/candidate_generator.py:1519]: Can we just add a small comment saying it can be 3d sometimes as well?

File: src/transformers/modeling_layers.py [hunk 4]
```diff
@@ -496,20 +496,28 @@ def forward(
...
+            else:
+                text_position_ids = rope_position_ids = position_ids
```
Comment (Cyrilvallez) [inline src/transformers/modeling_layers.py:516]: What a ternary! 🤯

File: tests/generation/test_utils.py (hunk 2 above)
Comment (Cyrilvallez) [reply tests/generation/test_utils.py]: As per my other comment, probably a good opportunity to start making a "subdir" of conversions per model_type for mtp!

**Review by Cyrilvallez — COMMENTED** [review]
Nice! As was noted already by @zucchini-nlp, and following something I discussed with you for Inkling, I think now is a good time to start adding a new level for the mtp conversions, as each main_model_type expects a different renaming/format for its mtp weights!

File: src/transformers/conversion_mapping.py
Comment (molbap) [reply src/transformers/conversion_mapping.py]: Alllright will look into it

File: src/transformers/modeling_layers.py (hunk 1 above)
Comment (molbap) [reply src/transformers/modeling_layers.py]: I'm using `mrope_section` to identify this for now

**Comment by molbap** [comment]
I went for a split within `conversion_mapping` @Cyrilvallez , LMK what you think!

**Comment by Cyrilvallez** [comment]
Ha damn, super super sorry @molbap this got a bit out of my stack, just seeing it again now 🥲 Let me know when you fix the conflicts, and I'll review again!
